spec.md

```diff
@@ -744,6 +744,202 @@ expose:
   - "8000"
 ```
 
+### extends
+
+Extend another service, in the current file or another, optionally overriding configuration. You can use
+`extends` on any service together with other configuration keys. The `extends` value must be a mapping
+defined with a required `service` and an optional `file` key.
+
+```yaml
+extends:
+  file: common.yml
+  service: webapp
+```
+
+If supported Compose implementations MUST process `extends` in the following way:
+
+- `service` defines the name of the service being referenced as a base, for example `web` or `database`.
+- `file` is the location of a Compose configuration file defining that service.
+
+#### Restrictions
+
+The following restrictions apply to the service being referenced:
+
+- Services with `links` cannot be used as a base
+- Services with `volumes_from` cannot be used as a base
+- Services with `net: container` cannot be used as a base
+- Services with `network_mode: service` cannot be used as a base
+- Services with `depends_on` cannot be used as a base
+- Services cannot have circular references with `extends`
+
+Compose implementations MUST return an error in all of these cases.
+
+#### Finding referenced service
+
+`file` value can be:
+
+- Not present.
+  This indicates that another service within the same Compose file is being referenced.
+- File path, which can be either:
```

Comment:
In other section we recommend not using relative path to enforce portability. Maybe we should introduce a dedicated section and just have references where relevant. I didn't know about `~user`, this sounds a bit hack-ish and highly unportable. I'm not sure we should have this being part of the spec (can be a docker-compose implementation detail)

Reply:
Removed the user path expansion bit (~ and ~user). I'm open to creating a separate section for relative paths 🙂